config_proc_macro/src/define_config.rs

```diff
@@ -0,0 +1,480 @@
+use std::convert::{From, TryFrom};
+
+use proc_macro2::TokenStream;
+use quote::{format_ident, quote, ToTokens};
+use thiserror::Error;
+
+pub fn define_rustfmt_config(input: syn::Item) -> TokenStream {
+    match input {
+        syn::Item::Struct(st) => define_rustfmt_config_on_struct(st),
+        _ => panic!("expected struct"),
+    }
+}
+
+fn define_rustfmt_config_on_struct(st: syn::ItemStruct) -> TokenStream {
+    let attrs = &st.attrs;
+    let vis = &st.vis;
+    let name = &st.ident;
+    let field_iter = st.fields.iter();
+    let non_config_fields = field_iter
+        .clone()
+        .filter(|f| f.attrs.iter().find(|a| is_rustfmt_config_attr(a)).is_none());
+    let config_defs: Vec<RustfmtConfigDef> = field_iter
+        .filter_map(|f| RustfmtConfigDef::try_from(f).ok())
+        .collect::<Vec<_>>();
+
+    let fields = config_defs.iter().map(RustfmtConfigDef::to_field);
+    let getters = config_defs.iter().map(RustfmtConfigDef::to_getter);
+    let setters = config_defs.iter().map(RustfmtConfigDef::to_setter);
+    let is_sets = config_defs.iter().map(RustfmtConfigDef::to_is_set);
+    let is_stables = config_defs.iter().map(RustfmtConfigDef::to_is_stable);
+    let validates = config_defs.iter().map(RustfmtConfigDef::to_validate);
+    let override_arms = config_defs.iter().map(RustfmtConfigDef::to_override_arm);
+    let is_valid_name_arms = config_defs
+        .iter()
+        .map(RustfmtConfigDef::to_is_valid_name_arm);
+    let is_default_arms = config_defs.iter().map(RustfmtConfigDef::to_is_default_arm);
+    let hash_set_inserts = config_defs.iter().map(RustfmtConfigDef::to_hash_set_insert);
+    let max_name_len = config_defs
+        .iter()
+        .map(|cfg_def| cfg_def.name.to_string().len())
+        .max()
+        .unwrap();
+    let print_docs = config_defs
+        .iter()
+        .map(|cfg_def| cfg_def.to_print_doc(max_name_len));
+    let explicit_defaults = config_defs
+        .iter()
+        .map(RustfmtConfigDef::to_explicit_default);
+
+    quote! {
+        #(#attrs)*
+        #vis struct #name {
+            #(#non_config_fields),*,
+            #(#fields),*
+        }
+
+        #[derive(Debug, thiserror::Error)]
+        pub enum ConfigError {
+            UnstableConfigOption(Vec<&'static str>),
+            TomlParseError(toml::de::Error),
+            TomlSerializationError(toml::ser::Error),
+            InvalidKeyValuePair(String, String),
+        }
+
+        impl std::fmt::Display for ConfigError {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+                match self {
+                    ConfigError::UnstableConfigOption(unstables) => {
```

Comment:
This will resolve #3872 correct?

Reply:
Unfortunately no, IIRC the issue is that we are parsing the same rustfmt.toml and emitting errors multiple times.